sympy/polys/matrices/domainmatrix.py

```diff
@@ -30,6 +30,29 @@ def __init__(self, rows, shape, domain):
         self.shape = shape
         self.domain = domain
 
+    def __getitem__(self, key):
+        m, n = self.shape
+        i, j = key
+        flag = False
+        if isinstance(i, slice):
+            flag = True
+            i, m = __getindex__(i, m)
+            m = m + 1
+        if isinstance(j, slice):
+            flag = True
+            j, n = __getindex__(j, n)
+            n = n + 1
+        if flag:
+            mat = []
+            for x in range(i, m):
+                list = []
+                for y in range(j, n):
+                    list.append(self.rep[x][y])
+                mat.append(list)
+            return DomainMatrix(mat, (m-i, n-j), self.domain)
```

Comment:
This logic should be handled in `DDM`. There should also be an implementation for `SDM` that handles sparse matrices efficiently (i.e. that does not loop over `range(i, m)`.

I think it would be best to add a method to both `DDM` and `SDM` like:
```
DDM.extract_slice(slice1, slice2)  -> DDM
SDM.extract_slice(slice1, slice2)  -> SDM
```
Then we can make an efficient implementation for each and this function can just do:
```python
return self.from_rep(self.rep.extract_slice(range(i, m), range(j, n))
```
The `Matrix` class already has an `extract` method and it would be similar to that but using slices because they can be used efficiently with `DDM`. Perhaps also there could be a method `extract` for `DDM` and `SDM` that works the same as for `Matrix`.

An efficient implementation for `DDM` would like:
```
def extract_slice(self, slice1, slice2):
    if not (isinstance(slice1, slice) and isinstance(slice2, slice)):
        raise TypeError('message')
    # Slicing lists is very efficient:
    elements = [row[slice2] for row in self[slice1]
    # I can't remember what the constructor looks like:
    return self.new(elements, self.domain)
```

Reply:
We have implemented for `SDM` and `DDM`.
For `SDM`, we are not using nested loops but introducing a new method `extract_map`.

@oscarbenjamin, Kindly review the latest commit.
Thanks.